File: Week 6 (Python)/sentimental-credit/credit.py
from cs50 import get_int


def main():
    card_num = get_int("Type the card number: ")
    if valid_luhn_card(card_num):
        get_card_type(card_num)
    else:
        print("INVALID")


def get_card_type(number):
    if int(number / 100000000000000) <= 55 and int(number / 100000000000000) > 50:
        print("MASTERCARD")
    elif int(number/1000000000000000) == 4 or int(number/1000000000000) == 4:
        print("VISA")
    elif int(number/10000000000000) == 34 or int(number/10000000000000) == 37:
        print("AMEX")
    else:
        print("INVALID")


def valid_luhn_card(card):
    sum = 0
    while (card > 0):
        last_digit = int(card % 10)
        card = int(card/10)
        dup_penult = int(card % 10) * 2
        if dup_penult > 9:
            firs_char = int(dup_penult % 10)
            dup_penult = int(dup_penult / 10)
            sum += dup_penult + firs_char
        else:
            sum += dup_penult
        sum += last_digit
        card = int(card / 10)
    if sum % 10 == 0:
        return True
    return False


main()

File: Week 1 (C)/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int check_sum(long long n);

int main(void)
{
    // Wait for input
    long long card_num = get_long("Digite o número do cartão: ");

    //Check if the is valid(with Luhn's Algorithm)
    int flag_sum = check_sum(card_num);



    //If is a Invalid (sum module ten != 0) then return INVALID msg
    if (flag_sum == 1)
    {
        printf("INVALID\n");
    }
    else
    {
        //Check the initial values are compatible by VISA or Mastercard or AMEX cards

        long long tentotwelve = 1000000000000;
        long long tentothirteen = 10000000000000;
        long long tentofourteen = 100000000000000;
        long long tentofifteen = 1000000000000000;
        if(card_num / tentofifteen == 4 || card_num / tentotwelve == 4)
        {
            printf("VISA\n");
        }
        else if(card_num/tentofourteen <=55 && card_num/tentofourteen > 50)
        {
            printf("MASTERCARD\n");
        }
        else if(card_num/tentothirteen == 34 || card_num/tentothirteen == 37)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
}

int check_sum(long long n)
{
    //This functions receive a credit card and verify if is a validy card number(with Luhn's algorithm)
    int sum = 0;
    while (n > 0)
    {
        //In each loop, the last 2 digits are taken, the second is multiplied and
        // added to the sum variable
        int last_digit = n % 10;
        n /= 10;
        int sec_lastdigit = n % 10;
        int dup_sec_lastdigit = sec_lastdigit *2;
        if (dup_sec_lastdigit > 9)
        {
            int firstintnum = dup_sec_lastdigit%10;
            dup_sec_lastdigit /= 10;
            sum += dup_sec_lastdigit + firstintnum + last_digit;
        }
        else
        {
            sum += (sec_lastdigit*2 + last_digit);
        }
        n /= 10;
    }
    //if the sum modulus is equal to 0, it returns 0 (it's valid card)
    if (sum % 10 == 0)
    {
        return 0;
    }
    else{
        return 1;
    }
}